microapps/FeatureFormsApp/app/src/main/java/com/arcgismaps/toolkit/featureformsapp/data/PortalItemRepository.kt

```diff
@@ -0,0 +1,152 @@
+
+    /**
+     * Refreshes the underlying data source to fetch the latest content. [forceUpdate] when set to
+     * true, will clear the existing cache.
+     *
+     * This operation is suspending and will wait until the underlying data source has finished
+     * AND the repository has finished loading the portal items.
+     */
+    suspend fun refresh(forceUpdate: Boolean = false) = withContext(dispatcher) {
+        if (forceUpdate) deleteAllCacheEntries()
+        portalItems.clear()
+        // get local items
+        val localItems = getListOfMaps().map { ItemData(it) }
+        // get network items
+        val remoteItems = remoteDataSource.fetchItemData()
+        // load the portal items and add them to cache
+        loadAndCachePortalItems(localItems + remoteItems)
+    }
+
+    /**
+     * Returns the number of items in the repository.
+     */
+    suspend fun getItemCount(): Int = withContext(dispatcher) {
+        return@withContext itemCacheDao.getCount()
+    }
+
+    /**
+     * Loads the list of [items] into loaded portal items and adds them to the Cache.
+     */
+    private suspend fun loadAndCachePortalItems(items: List<ItemData>) {
+        val entries = items.map { itemData ->
+            val portalItem = PortalItem(itemData.url)
+            portalItem.load()
+            val thumbnailUri = portalItem.thumbnail?.let { thumbnail ->
+                thumbnail.load()
+                thumbnail.image?.bitmap?.let { bitmap ->
+                    createThumbnail(
+                        portalItem.itemId,
+                        bitmap
+                    )
+                }
+            } ?: ""
+            ItemCacheEntry(
+                itemData.url,
+                portalItem.toJson(),
+                thumbnailUri,
+                portalItem.portal.url
+            )
+        }
+        // purge existing items and add the updated items
+        createCacheEntries(entries)
+    }
+
+    /**
+     * Deletes and inserts the list of [entries] using the [ItemCacheDao].
+     */
+    private suspend fun createCacheEntries(entries: List<ItemCacheEntry>) =
+        withContext(dispatcher) {
+            itemCacheDao.deleteAndInsert(entries)
+        }
+
+    /**
+     * Deletes all entries in the database using the [ItemCacheDao].
+     */
+    private suspend fun deleteAllCacheEntries() =
+        withContext(Dispatchers.IO) {
+            itemCacheDao.deleteAll()
+            val thumbsDir = File("$filesDir/thumbs")
+            if (thumbsDir.exists()) thumbsDir.deleteRecursively()
+        }
+
+    /**
+     * Creates a JPEG thumbnail using the [bitmap] with [name] filename in the local files
+     * directory and returns the absolute path to the file.
+     */
+    private suspend fun createThumbnail(name: String, bitmap: Bitmap): String =
+        withContext(Dispatchers.IO) {
+            val thumbsDir = File("$filesDir/thumbs")
+            if (!thumbsDir.exists()) thumbsDir.mkdir()
```

Comment:
make all necesary folders, not just the leaf folder.

```suggestion
            if (!thumbsDir.exists()) thumbsDir.mkdirs()
```